immich-app/immich
# feat(server): trash asset

Fixes: #739.
Fixes #937.
Fixes #2840.

#### Changes made:
- Made it possible to soft delete ( Trash ) assets by default instead of permanently deleting them. The trashed assets are accessible only in the Trash page and will not show up anywhere else.
- Restore / Permanently remove assets from the trash

https://github.com/immich-app/immich/assets/139912620/9655e973-d831-4a9e-907a-1712fa339021

- Admin setting to disable trash if not needed and an option to set the duration (in days) to keep the asset in trash. If the assets are not restored before the configured date, they will be permanently removed from Immich

<img width="1204" alt="image" src="https://github.com/immich-app/immich/assets/139912620/63f6d28d-2d6e-40be-a452-8f2ec654662f">

- Mobile app implementation for the trash view

Side bar | Trash page | Selection
:-: | :-: | :-: |
![Side-bar](https://github.com/immich-app/immich/assets/139912620/65909881-3a96-4abe-912d-615e54438604)|![Mobile-Trash](https://github.com/immich-app/immich/assets/139912620/a8283587-2e59-4a8c-8468-27640c985c80)|![Mobile-Trash-1](https://github.com/immich-app/immich/assets/139912620/c4bd10c5-3817-4e1b-93be-cc9af90ceef9)

- If trash is enabled and an asset is removed from the web, it will not be re-uploaded by the mobile app as long as the asset is still in trash. However, if the asset is removed from the trash, it will still be uploaded from the mobile app in the next backup sync. (This can be addressed by displaying an alert to the user asking them to remove the local asset as well. But that has to be handled as a separate PR independent of this one)
- Changes made in #4045 are cherry picked since trash enabled / disabled handling required server config / features in mobile app
- `DELETE /asset/` moved to domain in server

> [!WARNING]
>Deleting local assets still deletes them permanently from the mobile app. The following two features has to be added, possibly in future PRs:
>  - An option in mobile app to delete local only asset / both local and remote asset during deletion
 > - An option to trash local assets as well (Probably by moving the file to Immich app's internal directory)

#### How was this tested:
- Trash an asset and ensure that is it not permanently deleted and accessible from trash
- Trash an asset with trash disabled and ensure that it gets permanently removed
- Restore an asset from trash and ensure that it gets displayed in all pages
- Ensure that the asset is not returned in search results, faces, albums, favorites, etc. once trashed
- Ensure that the asset gets returned in search, etc., once restored
- Ensure audit logs are working when the asset is permanently removed from Immich
- Ensured duplicate uploads still detect assets from trash
- Ensured that the trash page is not accessible in Mobile / Web once trash is disabled

#### Todo
- [x] verify search still works
- [x] verify face matching ignores deleted faces
- [x] verify audit table still works
- [x] mobile implementation
- [x] empty trash bin button on the web
- [x] system config for trash can feature flag
- [x] system config for trash can duration (auto delete after x days)
- [x] cron job uses system config for processing trashed assets
- [x] show confirm when clicking delete action of trash is disabled.
- [x] hide trash can in side menu if trash is disabled
